third_party/amd/lib/TritonAMDGPUToLLVM/DotOpToLLVM/WMMA.cpp

```diff
@@ -0,0 +1,246 @@
+  auto bTensorTy = b.getType().cast<RankedTensorType>();
+  auto dTensorTy = d.getType().cast<RankedTensorType>();
+  auto elemTy = aTensorTy.getElementType();
+
+  auto aEncoding = aTensorTy.getEncoding().cast<DotOperandEncodingAttr>();
+  auto bEncoding = bTensorTy.getEncoding().cast<DotOperandEncodingAttr>();
+  int kWidth = aEncoding.getKWidth();
+
+  auto repA =
+      wmmaLayout.getWMMARepForOperands(aTensorTy.getShape(), elemTy, kWidth, 0);
+  auto repB =
+      wmmaLayout.getWMMARepForOperands(bTensorTy.getShape(), elemTy, kWidth, 1);
+
+  assert(repA[1] == repB[0]);
+
+  Value loadedA = adaptor.getA();
+  Value loadedB = adaptor.getB();
+  Value loadedC = adaptor.getC();
+  auto numRepM = repA[0];
+  auto numRepN = repB[1];
+  auto numRepK = repA[1];
+
+  ValueTable ha = getValuesFromDotOperandLayoutStruct(
+      rewriter, typeConverter, loadedA, numRepM, numRepK,
+      aTensorTy.getElementType(), loc);
+  ValueTable hb = getValuesFromDotOperandLayoutStruct(
+      rewriter, typeConverter, loadedB, numRepN, numRepK,
+      aTensorTy.getElementType(), loc);
+  auto dstElemTy = dTensorTy.getElementType();
+  auto fc = typeConverter->unpackLLElements(loc, loadedC, rewriter);
+
+  unsigned warpSize = triton::gpu::getWarpSize(wmmaLayout);
+  // TODO get rid of magic numbers
+  unsigned vgprElemWidth = 32;
+  unsigned paddedOutputElemSize =
+      vgprElemWidth / dstElemTy.getIntOrFloatBitWidth();
+  // compute number of output elements that each thread holds for one WMMA
+  // instruction.
+  auto elemsPerVec = mnkDim[0] * mnkDim[1] * paddedOutputElemSize / warpSize;
+  auto dElemsToStorePerThread = mnkDim[0] * mnkDim[1] / warpSize;
+  auto vecTy = vec_ty(dstElemTy, elemsPerVec);
+  for (int m = 0; m < numRepM; ++m) {
+    for (int n = 0; n < numRepN; ++n) {
+      Value acc = undef(vecTy);
+      for (unsigned v = 0; v < dElemsToStorePerThread; ++v) {
+        acc = insert_element(vecTy, acc,
+                             fc[m * numRepN * dElemsToStorePerThread +
+                                n * dElemsToStorePerThread + v],
+                             i32_val(v * paddedOutputElemSize));
+      }
+      for (size_t k = 0; k < numRepK; k++) {
+        acc = generateWMMAOp(rewriter, loc, wmmaInstrType, ha[{m, k}],
+                             hb[{n, k}], acc);
+      }
+      for (unsigned v = 0; v < dElemsToStorePerThread; ++v) {
+        fc[m * numRepN * dElemsToStorePerThread + n * dElemsToStorePerThread +
+           v] =
+            extract_element(dstElemTy, acc, i32_val(v * paddedOutputElemSize));
+      }
+    }
+  }
+
+  // replace with new packed result
+  Type structTy = LLVM::LLVMStructType::getLiteral(
+      wmmaLayout.getContext(), SmallVector<Type>(fc.size(), dstElemTy));
+  Value res = typeConverter->packLLElements(loc, fc, rewriter, structTy);
+  rewriter.replaceOp(op, res);
+  return success();
+}
+
+} // namespace
+
+LogicalResult convertWMMA(triton::DotOp op, triton::DotOp::Adaptor adaptor,
```

Comment:
can you add a basic lit tests to make sure this codes runs.

Reply:
Yep. working on it

Reply:
Need to merge https://github.com/openai/triton/pull/3171 first to create lit tests. Convert this PR to draft for now...